src/packet.rs

```diff
@@ -352,7 +358,59 @@ pub mod play {
         0x23 => BlockChange { location: BlockPos, block_id: Var<i32> }
         0x24 => BlockAction { location: BlockPos, byte1: u8, byte2: u8, block_type: Var<i32> }
         0x25 => BlockBreakAnimation { entity_id: Var<i32>, location: BlockPos, destroy_stage: i8 }
-        // 0x26 => MapChunkBulk { sky_light_sent: bool, chunks: Vec<Chunk>; impl Protocol for MapChunkBulk { ... } } // PROBLEM: chunks is encoded as two arrays, the first one specifying which sections of each chunk column are empty
+        0x26 => ChunkDataBulk { sky_light_sent: bool, chunk_meta: Vec<ChunkMeta>, chunk_data: Vec<ChunkColumn>;
+            impl Protocol for ChunkDataBulk {
+                type Clean = Self;
+                fn proto_len(this: &Self) -> usize {
+                    let mut len = 1; // sky_light_sent(bool) len is constant
+                    let columns = this.chunk_meta.len() as i32;
+                    len += <Var<i32> as Protocol>::proto_len(&columns);
+                    for cm in this.chunk_meta.iter() {
+                        len += <ChunkMeta as Protocol>::proto_len(cm);
+                    }
+                    for cd in this.chunk_data.iter() {
+                        len += cd.len();
+                    }
+                    len
+                }
+                fn proto_encode(this: &Self, mut dst: &mut Write) -> io::Result<()> {
+                    try!(<bool as Protocol>::proto_encode(&this.sky_light_sent, dst));
+                    let columns = this.chunk_meta.len() as i32;
+                    try!(<Var<i32> as Protocol>::proto_encode(&columns, dst));
+                    for cm in this.chunk_meta.iter() {
+                        try!(<ChunkMeta as Protocol>::proto_encode(cm, dst));
+                    }
+                    for cd in this.chunk_data.iter() {
+                        let chunk_column = try!(cd.encode());
+                        try!(dst.write_all(&chunk_column));
+                    }
+                    Ok(())
+                }
+                fn proto_decode(mut src: &mut Read) -> io::Result<ChunkDataBulk> {
+                    let sky_light_sent = try!(<bool as Protocol>::proto_decode(src));
+                    let columns = try!(<Var<i32> as Protocol>::proto_decode(src));
+                    let mut chunk_meta = Vec::with_capacity(columns as usize);
+                    for cm in chunk_meta.iter_mut() {
+                        *cm = try!(<ChunkMeta as Protocol>::proto_decode(src));
+                    }
+                    // Read all encoded ChunkColumns
+                    let mut data = Vec::with_capacity(1 << 12);
+                    try!(src.read_to_end(&mut data));
+                    let mut chunk_data = Vec::with_capacity(columns as usize);
+                    for (idx, cd) in chunk_data.iter_mut().enumerate() {
```

Comment:
I thought zipping both loops in just one. Your proposal is better yes, applying...